Added comparison functions (GT/LT) to dice: !roll 1d20 + x > y

NicholasClaaszen opened the pull request:
critique welcome, as always. 

I think checkTarget() can be streamlined, but I'm not sure how

NicholasClaaszen committed:
Added comparison functions (GT/LT) to dice: !roll 1d20 + x > y

NicholasClaaszen committed:
Accidentally left in some debug code

ryanguill commented on plugins/dice.js:
```diff
@@ -3,9 +3,29 @@ const _ = require('lodash');
+			} else {
+				text = ' Failure! ' + result + ' is higher than ' + target;
+			}
+		} else {
```
could the quantifier be equal or not equal to? or what about GTE/LTE?

ryanguill commented on plugins/dice.js:
```diff
@@ -36,12 +56,23 @@ module.exports = (function(){
+			let [diePart, quantifier, target] = element.split(/(&[g|G|l|L]t;)/g),
+				[die, operator, modifier] = diePart.split(/(\+|-|\*|\/)/g),
 				roll = rollDice(die);
+			console.log({
```
dont forget to change this to bot.log() or remove

ryanguill commented on plugins/dice.js:
```diff
@@ -3,9 +3,29 @@ const _ = require('lodash');
 
+	function checkTarget(result, quantifier, target) {
+		let text = '';
+		if (quantifier === '&gt;') {
```
would a switch() work better than the ifs? youre only ever checking quantifier right?

ryanguill commented on plugins/dice.js:
```diff
@@ -36,12 +56,23 @@ module.exports = (function(){
 			if (operator) roll.result = diceCalculation(roll.result, operator, modifier);
-			returnMessage = returnMessage + ' ' + element + ' = ' + roll.result + ' [' + roll.rolls.toString() + ']';
+			if (quantifier && target) roll.target = checkTarget(roll.result, quantifier, target);
+			roll.target = roll.target || '';
```
if you swapped 72 and 73 youd get rid of the ||, just set to blank and if quantifier and target overwrite it. or get rid of this line and do the || when concatenating on 75.

ryanguill commented on plugins/dice.js:
```diff
@@ -3,9 +3,29 @@ const _ = require('lodash');
+		reDice = /([1-9][0-9]{0,2})(d|D)([1-9][0-9]{0,3})((\+|-|\*|\/)([0-9]{1,3}))?( *\&(g|G|l|L)t\; *[0-9]*)?/g,
 		reMessage = /(^(\!roll )[1-9][0-9]{0,2})(d|D)([1-9][0-9]{0,3})((\+|-|\*|\/)([0-9]{1,3}))?/g;
 
+	function checkTarget(result, quantifier, target) {
```
I think you could simplify a lot if you just showed pass or fail and then built a comparison function that took the operator and result and target.  but I dont think what you have is too complicated.

NicholasClaaszen committed:
checkTarget & Regex
- Changed the checkTarget function to use different, cooler syntax.
- Included E, GTE and LTE checks.

NicholasClaaszen commented:
Updated pull request with suggestions.

NicholasClaaszen committed:
Changed mark-up to Codeblocks

NicholasClaaszen committed:
Changed mark-up to match Adam Tuttle's suggestions.

ryanguill commented on plugins/dice.js:
```diff
@@ -3,8 +3,21 @@ const _ = require('lodash');
+			};
+		if (quantifiers[quantifier.toLowerCase()](parseFloat(result), parseFloat(target))) text = ' `Success!`';
+		return text;
+	}
```
this is getting nitpicky, there really isnt anything wrong with what you have (except for maybe the parsefloat()) but here is how I would write this function:

``` js
function checkTarget(result, quantifier, target) {

    //do your argument manipulation, defaulting, etc at the top of the function
    //making your use of them more clear
    quantifier = quantifier.toLowerCase();
    //either use lodash's parseFloat or use parseFload(result, 10)
    result = _.parseFloat(result);
    target = _.parseFloat(target);

    let quantifiers = {
        //personal style preferences
        '&gt;': (a, b) => a > b,
        '&lt;': (a, b) => a < b,
        '&gt;=': (a, b) => a >= b,
        '&lt;=': (a, b) => a <= b,
        '=': (a, b) => a === b
    };

    //this line is much easier to read with the manipulation stuff already done
    if (quantifiers[quantifier](result, target)){
        //I personally only use bracket-less if statements when doing an early bail-out return,
        //like when the function should just not run - errors or something like that
        return ' `Success!`';
    }
    //no need to store the result, just return it where necessary
    return ' `Failure!`'
}
```

ryanguill replied:
more info on parseInt() http://stackoverflow.com/questions/16880327/why-am-i-getting-weird-result-using-parseint-in-node-js-different-result-from

you can also use `Number()` in most cases instead of parseInt

ryanguill commented on plugins/dice.js:
```diff
@@ -12,7 +25,9 @@ module.exports = (function(){
 			let roll = Math.floor((Math.random() * sides) + 1);
-			result = result + roll;
+			result += roll;
+			if (roll === 1) roll = '`' + roll + '`';
```
similar to above, I try to avoid bracket-less if statements except in very small circumstances.  Here is some reasoning: http://stackoverflow.com/questions/4797286/are-curly-braces-necessary-in-one-line-statements-in-javascript

ryanguill commented on plugins/dice.js:
```diff
@@ -34,14 +49,16 @@ module.exports = (function(){
 		for (let element of aDiceRolls) {
-			let [die, operator, modifier] = element.split(/(\+|-|\*|\/)/g),
+			let [diePart, quantifier, target] = element.split(/(&[g|l]t;=*|=)/g),
+				[die, operator, modifier] = diePart.split(/(\+|-|\*|\/)/g),
```
:+1: for the destructuring, works well here!

ryanguill commented on plugins/dice.js:
```diff
@@ -34,14 +49,16 @@ module.exports = (function(){
+			roll.target = '';
+			if (quantifier && target) roll.target = checkTarget(roll.result, quantifier, target);
+			returnMessage += '( ' + element + ' ) = `' + roll.result + '` [' + roll.rolls.toString() +
+				']' + roll.target + ' ';
```
try es6 template strings here:
(not tested)

``` js
returnMessage += `(${element}) = \`${roll.result}\` [${roll.rolls.toString()}]${roll.target} `;
```

atuttle commented:
If/when this lands it will require updates to the `!roll` docs added in #52
